src/workerd/api/deferred-proxy-test.c++

```diff
@@ -0,0 +1,233 @@
+#include "deferred-proxy.h"
+#include <kj/test.h>
+
+namespace workerd::api {
+namespace {
+
+KJ_TEST("kj::Promise<DeferredProxy<T>>: early co_return implicitly fulfills outer promise") {
+  kj::EventLoop loop;
+  kj::WaitScope waitScope(loop);
+
+  {
+    // Implicit void co_return.
+    auto coro = []() -> kj::Promise<DeferredProxy<void>> {
+      co_await kj::Promise<void>(kj::READY_NOW);
+    };
+    auto promise = coro();
+    KJ_EXPECT(promise.poll(waitScope));
+    auto proxyTask = promise.wait(waitScope).proxyTask;
+    KJ_EXPECT(proxyTask.poll(waitScope));
+    proxyTask.wait(waitScope);
+  }
+  {
+    // Explicit void co_return.
+    auto coro = []() -> kj::Promise<DeferredProxy<void>> {
+      co_return;
+    };
+    auto promise = coro();
+    KJ_EXPECT(promise.poll(waitScope));
+    auto proxyTask = promise.wait(waitScope).proxyTask;
+    KJ_EXPECT(proxyTask.poll(waitScope));
+    proxyTask.wait(waitScope);
+  }
+  {
+    // Valueful co_return.
+    auto coro = []() -> kj::Promise<DeferredProxy<int>> {
+      co_return 123;
+    };
+    auto promise = coro();
+    KJ_EXPECT(promise.poll(waitScope));
+    auto proxyTask = promise.wait(waitScope).proxyTask;
+    KJ_EXPECT(proxyTask.poll(waitScope));
+    KJ_EXPECT(proxyTask.wait(waitScope) == 123);
+  }
+}
+
+KJ_TEST("kj::Promise<DeferredProxy<T>>: `KJ_CO_MAGIC BEGIN_DEFERRED_PROXYING` fulfills outer "
+    "promise") {
+  kj::EventLoop loop;
+  kj::WaitScope waitScope(loop);
+
+  auto paf1 = kj::newPromiseAndFulfiller<void>();
+  auto paf2 = kj::newPromiseAndFulfiller<int>();
+
+  auto coro = [&]() -> kj::Promise<DeferredProxy<int>> {
+    co_await paf1.promise;
+    KJ_CO_MAGIC BEGIN_DEFERRED_PROXYING;
```

Comment:
Looking at this, if we're going to define a kj macro for this, we may as well define it for the entire line... e.g.

e.g.
```
KJ_CO_BEGIN_DEFERRED_PROXYING;
```

Or do we imagine that `KJ_CO_MAGIC` will be used for other cases?

Reply:
Yes, we do: https://github.com/capnproto/capnproto/pull/1739